### PR TITLE
Remove diagnostics from error list on document close or reset for tha…

### LINKED ISSUE #39902: Error list experience on closing documents
Follow-up issue to discuss https://github.com/dotnet/roslyn/pull/39892#issue-342926809. That PR restores the prior error list behavior of removing closed file diagnostics from error list whenever a document is closed. This issue tracks discussing if we want to change that experience.

### PR DESCRIPTION
**Pull request by mavasani**
…t document

This fixes a recent regression where we stopped clearing out documents diagnostics on document close for TypeScript.

Fixes #39902

--------------------
**More Context**

1. Prior behavior (when closed file diagnostics/FSA is turned off):
    1. We only compute diagnostics for opened files and populate error list
    2. When user closes a file, we explicitly remove the diagnostics from the error list for that file.
2. Recently, we added [new functionality](https://github.com/dotnet/roslyn/pull/39596) to re-purpose “Run Code Analysis” command for C# and VB projects, which regressed the above behavior.
    1. With this command, users can force compute live diagnostics for any project or current solution on demand. Computed diagnostics include both open file and closed file diagnostics and they are automatically populated in the error list.
    2. We want these diagnostics to stay in the error list even for closed files. I had to make a change to diagnostic service to not force remove all closed file diagnostics, otherwise this command was populating the correct diagnostics in error list only for them to be immediately removed for all closed files.
3. New behavior:
    1. We only compute diagnostics for opened files and populate error list by default
    2. Users can force complete diagnostics for closed files by using explicit commands, at which point error list will also contain closed file diagnostics
    3. When user closes the file, we do not remove the diagnostics from the error list. User can switch the error list scope combo box to “Open documents” instead of the default “Entire solution” to only view diagnostics for opened documents.

We feel the new behavior is more logical – error list contents are purely what diagnostics have been computed and is not affected by actions such as closing a file. Users should be taught to use the error list scope combo box if they desire to get differing views over data, such as Current Document, Open Documents, and so on, instead of us trying to proactively do so for users by modifying the actual error list contents.

However, TypeScript would like to retain the old behavior. So we have decided to add a hook to allow a per-language setting to decide between the old and new behavior.

### CODE REVIEW

**Comment by amcasey** [comment]
Thanks!  FYI @mrcrane

**Comment by mavasani** [comment]
FYI: I am working on an integration test for this behavior to prevent future regressions.

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn/issues/39902 to discuss https://github.com/dotnet/roslyn/pull/39892#issue-342926809 in more detail at the IDE design meeting.

**Comment by mavasani** [comment]
@sharwell - I have pushed the changes so the new behavior is retained for C# and VB, and old behavior is restored for TypeScript. I have temporarily hardcoded TypeScript for the old behavior, but that can be removed once they move to using the new option to enable the old behavior.

Tagging @jinujoseph

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs [hunk 1]
```diff
@@ -196,21 +220,26 @@ public void RemoveDocument(DocumentId documentId)
...
+                RaiseDiagnosticsRemovedForDocument(documentId, stateSets);
             }
         }
 
+        private void RaiseDiagnosticsRemovedForDocument(DocumentId documentId, IEnumerable<StateSet> stateSets)
+        {
+            // remove all diagnostics for the document
+            Owner.RaiseBulkDiagnosticsUpdated(raiseEvents =>
+            {
+                Solution nullSolution = null;
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs]: 💡 This can be inlined as `solution: null`

**Comment by minestarks** [comment]
@mavasani  I reran our tests after installing the VSIX from this CI, but they still fail with the same behavior. I haven't had a chance to dig into why, but will follow up after the holidays.

**Comment by mavasani** [comment]
@minestarks I verified that by replacing the hard coded TypeScript string with C# ensures that closed file diagnostics for C# go away. I am wondering if the TS are failing due to some other reason. Would it be possible to manually validate your scenario?

**Comment by mavasani** [comment]
@minestarks I manually validated the TS scenario (as per my understanding). I opened a new TS file, which creates a virtual TS project with bunch of library files. I opened one of those library files, then introduced syntax errors and closed the file. The errors remain in error list after closing the library file on latest dogfood build, but they are removed from error list after applying changes on this branch.

**Comment by mavasani** [comment]
@minestarks Let me know if you were able to manually validate this.

**Comment by minestarks** [comment]
@mavasani  thanks for following up. It's entirely possible there was an issue with the test set up. For what it's worth, the test opens an asp.net project and adds a .ts file to it (so not just a loose .ts file). I will run the steps manually once I have access to a machine again (tomorrow)

https://devdiv.visualstudio.com/DevDiv/_git/TypeScript-VS?path=%2FVS%2FLanguageService%2FTests%2FIntegration%2FClosedFileDiagnostics%2FClosedFileDiagnosticsTests.cs&version=GBmaster&line=13&lineEnd=15&lineStartColumn=1&lineEndColumn=113&lineStyle=plain

**Comment by minestarks** [comment]
@mavasani had a chance to try the test manually, it worked. :shipit: Hopefully it was something in my set up and the next Roslyn insertion will turn the integration test green as well. Thanks!

**Comment by mavasani** [comment]
Thanks @minestarks!

**Pull request merged**